external-crates/move/crates/move-compiler/src/parser/ast.rs

```diff
@@ -72,11 +72,29 @@ pub struct Program {
     pub lib_definitions: Vec<PackageDefinition>,
 }
 
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum ExternalTargetKind {
+    Library,
+    SkippedSource,
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+/// Specifies a source target or dependency
+pub enum TargetKind {
+    /// A source module. If is_root_package is false, some warnings might be suppressed.
+    /// Bytecode/CompiledModules will be generated for any Source target
+    Source { is_root_package: bool },
+    /// A dependency only used for linking.
+    /// No bytecode or CompiledModules are generated
+    External(ExternalTargetKind),
+}
```

Comment:
```
pub enum TargetSource {
    Root,
    Dependency,
}

pub enum TargetKind {
    Compiled(TargetSource),
    External(TargetSource),
}
```

Reply:
Final bike shedding... hopefully. I think this `enum` is what is going on?

But at this point maybe at that point we should just have 
```
public struct TargetKind {
    is_root_package: bool,
    has_bytecode_generated: bool,    
}
```
?

Reply:
I think it's a little more subtle than that but I may be wrong. From what I understand, `Source` and `External` in `TargetKind` are mutually exclusive so making `TargetKing` into a struct would be a bit misleading.

The `Source` variant covers all source files that are fully compiled (i.e., their function bodies are compiled and their bytecode generated)), whether they belong to dependencies (where sources are provided) or to the root package.

The `External` variant covers cases where no function bodies are compiled (no bytecode or ASTs beyond parsing are available for them), and covers two cases:
- dependency (library) provided without a source to the compiler (as per `ModuleFormat` [enum](https://github.com/MystenLabs/sui/blob/1ae9af6963d2a61e06be3ca3fc82ca94cfe08b68/external-crates/move/crates/move-package/src/compilation/compiled_package.rs#L121))
- explicitly skipped function body compilation for source files specified at compiler invocation (the new stuff)